tests/unit/test_library_smtp.py

```diff
@@ -0,0 +1,272 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""SMTP library unit tests"""
+import ops
+import pytest
+from charms.smtp_integrator.v0 import smtp
+from ops.testing import Harness
+
+REQUIRER_METADATA = """
+name: smtp-consumer
+requires:
+  smtp:
+    interface: smtp
+  smtp-legacy:
+    interface: smtp
+"""
+
+PROVIDER_METADATA = """
+name: smtp-producer
+provides:
+  smtp:
+    interface: smtp
+  smtp-legacy:
+    interface: smtp
+"""
+
+
+class SmtpRequirerCharm(ops.CharmBase):
+    """Class for requirer charm testing."""
+
+    def __init__(self, *args):
+        """Init method for the class.
+
+        Args:
+            args: Variable list of positional arguments passed to the parent constructor.
+        """
+        super().__init__(*args)
+        self.smtp = smtp.SmtpRequires(self)
+        self.smtp_legacy = smtp.SmtpRequires(self, relation_name=smtp.LEGACY_RELATION_NAME)
+        self.events = []
+        self.framework.observe(self.smtp.on.smtp_data_available, self._record_event)
+        self.framework.observe(self.smtp_legacy.on.smtp_data_available, self._record_event)
+
+    def _record_event(self, event: ops.EventBase) -> None:
+        """Record emitted event in the event list.
+
+        Args:
+            event: event.
+        """
+        self.events.append(event)
+
+
+class SmtpProviderCharm(ops.CharmBase):
+    """Class for provider charm testing."""
+
+    def __init__(self, *args):
+        """Init method for the class.
+
+        Args:
+            args: Variable list of positional arguments passed to the parent constructor.
+        """
+        super().__init__(*args)
+        self.smtp = smtp.SmtpProvides(self)
+        self.smtp_legacy = smtp.SmtpProvides(self, relation_name=smtp.LEGACY_RELATION_NAME)
+        self.events = []
+        self.framework.observe(self.on.smtp_relation_changed, self._record_event)
+        self.framework.observe(self.on.smtp_legacy_relation_changed, self._record_event)
+
+    def _record_event(self, event: ops.EventBase) -> None:
+        """Record emitted event in the event list.
+
+        Args:
+            event: event.
+        """
+        self.events.append(event)
+
+
+def test_smtp_provider_charm_relations():
```

Comment:
Do we need this test? It seems to be testing Juju's `add_relation` rather than the provider side.

Reply:
The provider logic is just to execute the appropriate *_relation_changed handler, that's what I'm actually asserting at the end

Reply:
From what I see in the code, the Provider side has no relation to any events. It has one public property (discussed in another question) and one public method for updating the relation data bag. 
Also the test doesn't really assert that the appropriate handler has been executed. It asserts the number of relations of certain name in the model, so in fact it's just testing Juju. 
Perhaps `SmtpProvides` should be handling relation_joined events rather than the charm?

Reply:
Sorry, just forget what I said, I got confused. This tests the `relations` public property of the `SmtpProvides` class of the lib. Maybe it is a bit unclear because that class is defined an an attribute of the charm `SmtpProviderCharm` I use for testing

Reply:
Addressed in #27